File: sharedmodule/src/main/java/com/workingbit/share/domain/impl/Square.java
package com.workingbit.share.domain.impl;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.workingbit.share.domain.BaseDomain;
import com.workingbit.share.domain.ICoordinates;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Aleksey Popryaduhin on 09:26 10/08/2017.
 */
@Data
@AllArgsConstructor
@NoArgsConstructor
@EqualsAndHashCode(of = {"v", "h"})
public class Square implements ICoordinates, BaseDomain {

  /**
   * row
   */
  private int v;
  /**
   * col
   */
  private int h;

  /**
   * Board's dimension
   */
  @JsonIgnore
  private int dim;

  /**
   * on the main part where we have the draughts
   */
  private boolean main;

  /**
   * if square highlighted for allowing to move
   */
  private boolean highlighted;

  private Draught draught;

  @JsonIgnore
  private List<List<Square>> diagonals = new ArrayList<>();

  public Square(int v, int h, int dim, boolean main, Draught draught) {
    this.v = v;
    this.h = h;
    this.dim = dim;
    this.main = main;
    this.draught = draught;
  }

  public Square(int v, int h, int dim, boolean main) {
    this(v, h, dim, main, null);
  }

  public boolean isOccupied() {
    return draught != null;
  }

  @Override
  public String toString() {
    return "Square{" +
        "notation=" + toNotation() +
        ", highlighted=" + highlighted +
        ", draught=" + draught +
        '}';
  }

  public void addDiagonal(List<Square> diagonal) {
    this.diagonals.add(diagonal);
  }
}

File: coremodule/src/main/java/com/workingbit/coremodule/domain/impl/Square.kt
package com.workingbit.coremodule.domain.impl

import com.fasterxml.jackson.annotation.JsonIgnore

/**
 * Created by Aleksey Popryaduhin on 17:03 23/08/2017.
 */
data class Square(private val map: HashMap<String, Any?>) {

    /**
     * row
     */
    var v: Int by map
    /**
     * col
     */
    var h: Int by map
    /**
     * on the main part where we have the draughts
     */
    val main: Boolean by map
    /**
     * if square highlighted for allowing to move
     */
    var highlighted: Boolean  by map
    val size: Int by map

    var draught: Draught? by map

    /**
     * Selected draught is point for new recursion
     */
    var pointDraught: Draught? by map
        @JsonIgnore get

    val occupied: Boolean
        get() = draught != null

    constructor(v: Int, h: Int, size: Int?, draught: Draught?) : this(
            hashMapOf("v" to v,
                    "h" to h,
                    "main" to ((h + v + 1) % 2 == 0),
                    "size" to size,
                    "draught" to draught))
}